README.md
stats\_batch
================
Christopher Gandrud

[![pytest-check](https://github.com/christophergandrud/stats_batch/actions/workflows/test-stats-batch.yaml/badge.svg)](https://github.com/christophergandrud/stats_batch/actions)
[![code-coverage](img/coverage.svg)](.coverage)

**In Development**

**stats\_batch** is a Python package for finding approximate statistics
using batch updating. Currently it supports:

- mean

- variance

- sum of squared deviations

- sample size

## Motivation

Imagine we have a stream of data and we want to make some descriptive
and statistical inferences with it. Rather than storing all of the data
and calculating statistics all at once, we can update the statistics as
new batches of data come in.

<img src="img/batch-update-basic-illustrate.png" width="50%" align="center"/>

Batch updating has at least two advantages:

-   We don’t need to store all of the data, only a small number of
    sufficient statatistics. This makes the calculations very fast.

-   Because we don’t need to store all of the data, we can minimise the
    amount of potentially sensitive personal data we need to store.

The key downside, is that due to rounding error, batch updated
statistics are not exact. Though the rounding error is typically not
substantively large (see [here](https://elegant-heyrovsky-54a43f.netlify.app/privacy-first-ds-mean-var.html)).

See
[here](https://elegant-heyrovsky-54a43f.netlify.app/privacy-first-ds-mean-var.html)
for more details on the motivation, algorithms, and sources.

## Install

Currently, you can install stats-batch from Test PyPi:

```bash
pip3 install -i https://test.pypi.org/simple/ stats-batch
```

## Example

### Minimal example: mean and variance

Currently, **stats_batch** supports updating the mean and variance of an array of numbers using `sb.mean_var_batch` and the `update` method. 
Here is a minimal example

``` python
import numpy as np
import stats_batch as sb

# Simulate data
n = 10_000
x = np.random.normal(size=n)

# First batch
batch_1_a = a[:100]  
a_current = sb.mean_var_batch(batch_1_a)

# Update with inputs from a second batch
batch_2_a = a[100:n]
a_current.update(batch_2_a)

a_current.print()
##    mean   var  sum_squared_dev  sample_size
## 0   2.5  1.25              5.0            4
```

### Output

You can output the statistics using the following methods:

- `print()`

- `to_pandas()`

- `to_csv(filename)`

For example:

```python
sb.mean_var_batch([1,2,3,4]).to_pandas()
##    mean   var  sum_squared_dev  sample_size
## 0   2.5  1.25              5.0            4
```

### Compare difference of means of two samples

Imagine we have an A/B test and want to ultimately calculate the
difference of means of the A and B groups.

First, for the sake of this example let’s simulate the data and create a
function that allows us to pretend the data is coming in as batches:

```python
import stats_batch.mean_var_batch as sb
import numpy as np
import stats_batch as sb
from scipy.stats import ttest_ind
from itertools import islice

# ----------------------------------------------------------------
"""
Simulate batched A/B test data
"""
# Simulate full samples
n = 10_000
batch_n = 1_000
a = np.random.normal(size=n, loc=0.1)
b = np.random.normal(size=n, loc=0.01)
```

Now let's batch update the data as it "comes in".

```python
"""
Batch update sufficient stattistics
"""
# Group A
for i, new_list in enumerate(sb.group_elements(a , batch_n)):
    if i == 0:
        mean_var_a = sb.mean_var_batch(new_list)
    else:
        mean_var_a.update(new_list)

# Group B
for i, new_list in enumerate(sb.group_elements(b , batch_n)):
    if i == 0:
        mean_var_b = sb.mean_var_batch(new_list)
    else:
        mean_var_b.update(new_list)
```

Finally, we run a t-test on the batch updated data. 

```python
# Compare means of A and B using batch updated sufficient statistics

mean_var_a.ttest_ind(mean_var_b)

## Ttest_indResult(statistic=6.541321847403076, pvalue=6.245323161218439e-11)
```

Notice that the t-test results from the batch updated data are very similar to the t-test results using the full samples, without batch updating.

```python
# Compare to t-test using full samples
ttest_ind(a, b)

## Ttest_indResult(statistic=6.541321847403074, pvalue=6.245323161218484e-11)
```

## Ambition

Version 0.1 of **stats_batch** provides the sufficient statistics needed to analyse the difference of means of two samples using a t-test. The near-ish term ambition of the package is to enable:

- enable Bayesian updating for inference as new data comes in

- enable regression adjustment through recursive least squares (e.g. [Harvey 1990](https://mitpress.mit.edu/books/econometric-analysis-time-series-second-edition)). This could, for example, be used for variance reduction in the analysis of A/B tests ([Deng et al.'s (2013)](https://exp-platform.com/Documents/2013-02-CUPED-ImprovingSensitivityOfControlledExperiments.pdf) CUPED method). 

## Sources

- Chou (2021) "Randomized Control Trials with Minimal Data Retention" <https://arxiv.org/abs/2102.03316>.

- Chan et al. (1983). "Algorithms for computing the sample variance: Analysis and recommendations". <http://www.cs.yale.edu/publications/techreports/tr222.pdf>

Note that there appears to be a typo in Chou's paper (as of March 2021) for the variance formula. It is corrected in `stats_batch` using Chane et al. (1983).


stats_batch/CHANGELOG
# All changes to batchstats are documented here

## Version 0.1

- First release version of batchstats. Includes `batch_mean`, `batch_var`, and the wrapper
of these two `batch_mean_var` as well as methods for exporting to Pandas and running
t-tests comparing means of two batch updated independent samples.

test/test_batch_var.py
import stats_batch as sb
import numpy as np
import numpy.testing as npt

# Test that var_batch returns exact `np.var` variance if there is 
# no prior batch information
def test_var_batch_missing_prior():
    x = list(range(1, 100))
    assert sb.var_batch(x)[0] == np.var(x)
    assert sb.var_batch(x)[1] == sb.sum_square_deviations(x)  

# Test that var_batch returns approximate `np.var` for batch updates
def test_var_batch_multiple_batches():
    n = 10_000
    x = np.random.normal(size=n)

    # First batch
    batch_1 = x[:100]  
    b1_mean, b1_n = sb.mean_batch(batch_1)
    b1_var, b1_ssd = sb.var_batch(batch_1) 

    # Second batch
    batch_2 = x[100:n]
    b2_var, b2_ssd = sb.var_batch(batch_2, b1_mean, b1_ssd, b1_n)
    npt.assert_approx_equal(b2_var, np.var(x), significant=4)
    npt.assert_approx_equal(b2_ssd, sb.sum_square_deviations(x))


test/test_batch_mean.py
import stats_batch as sb
import numpy as np

# Test mean_batch returns the mean if prior_mean and prior_sample_size are missing
def test_mean_batch_missing_prior_mean_prior_sample_size():
    x = list(range(1, 100))
    assert sb.mean_batch(x) == np.mean(x) 

tests/test_batch_mean.py
import stats_batch as sb
import numpy as np
import numpy.testing as npt

# Test mean_batch returns the mean if prior_mean and prior_sample_size are missing
def test_mean_batch_missing_prior_mean_prior_sample_size():
    x = list(range(1, 100))
    assert sb.mean_batch(x)[0] == np.mean(x) 
    assert sb.mean_batch(x)[1] == len(x)

# Test mean_batch returns the correct mean from multiple batches
def test_mean_batch_multiple_batches():
    n = 10_000
    x = np.random.normal(size=n)

    # First batch
    batch_1 = x[:100]  
    b1_mean, b1_n = sb.mean_batch(batch_1) 

    # Second batch
    batch_2 = x[100:n]
    b2_mean, b2_n = sb.mean_batch(batch_2, b1_mean, b1_n)
    npt.assert_approx_equal(b2_mean, np.mean(x))
    assert b2_n == n


stats_batch/batchmean.py
"""
Functions to use batch algorithms to find the mean and variance of a sample.
"""

import numpy as np


def batch_mean(new_batch, prior_sample_size:int=None, prior_mean:float=None) -> float:
    """
    Find the new mean of a sample updated by one batch. If only `new_batch` is 
    given, then `np.mean` is used.

    Parameters
    ----------
        new_batch: List[Union[int, float]]  
            List of all values in the new batch
        prior_sample_size: int
            Number of samples in the prior batches.
        prior_mean: float
            Mean up until the new batch.

    Returns
    -------
        The mean including the new batch and mean of the prior batches or, if only `new_batch` is given, the
        mean of the batch without updating.

    Examples
    -------- 
    >>> batch_mean([1,2,3,4])
    2.5
    >>> batch_mean([1,2,3,4], 4, 0)
    1.25

    Sources
    -------
    See Gandrud (2021) <https://elegant-heyrovsky-54a43f.netlify.app/privacy-first-ds-mean-var.html> for algorithm details. 
    """
    if prior_sample_size is None or prior_mean is None:
        return np.mean(new_batch)
    else:
        sum_new_batch = sum(new_batch)
        total_samples = prior_sample_size + len(new_batch)
        return prior_mean + (1/total_samples) * (sum_new_batch - ((total_samples - prior_sample_size) * prior_mean))




stats_batch/mean_var_batch.py
"""
Functions to use batch algorithms to find the mean and variance of a sample.
"""

import numpy as np
from scipy.stats import ttest_ind_from_stats
import pandas as pd
from math import sqrt

def sum_square_deviations(x) -> float:
    """
    Sum of the squared deviations of a sample from the mean.

    Parameters
    ----------
    x : array_like
        Sample to find the sum of the squared deviations of.

    Returns
    -------
    float 
        sum of the squared deviations of the sample from the mean.

    Examples
    --------
    >>> from stats_batch import sum_square_deviations
    >>> sum_square_deviations([1, 2, 3])
    2.0
    """
    return sum((x - np.mean(x)) ** 2)


def mean_batch(new_batch, prior_mean:float=None, prior_sample_size:int=None):
    """
    Find the new (approximate) mean of a sample updated by one batch. 
    If only `new_batch` is given, then `np.mean` is used.

    Parameters
    ----------
        new_batch: List[Union[int, float]]  
            List of all values in the new batch
        prior_mean: float
            Mean up until the new batch.
        prior_sample_size: int
            Number of samples in the prior batches.

    Returns
    -------
        tuple(float, int)
        1. The mean from the new batch and the mean of the prior batches. This will be *approximately* 
        equal to the total sample mean.
        If only `new_batch` is given, the mean of the batch is returned with the total sample size without updating.
        2. Total sample size. 

    Examples
    -------- 
    >>> from stats_batch import mean_batch
    >>> mean_batch([1,2,3,4])
    (2.5, 4)
    >>> mean_batch([1,2,3,4], prior_mean = 0, prior_sample_size = 4, )
    (1.25, 8)

    References
    ----------
    See Gandrud (2021) <https://elegant-heyrovsky-54a43f.netlify.app/privacy-first-ds-mean-var.html> for algorithm details.

    This work is based on:
    - Chou (2021, 5) <https://arxiv.org/pdf/2102.03316.pdf>

    - Chan et al (1983) <http://www.cs.yale.edu/publications/techreports/tr222.pdf> 
    """
    if prior_sample_size is None or prior_mean is None:
        return np.mean(new_batch), len(new_batch)
    else:
        sum_new_batch = sum(new_batch)
        total_samples = prior_sample_size + len(new_batch)
        updated_mean = prior_mean + (1/total_samples) * (sum_new_batch - ((total_samples - prior_sample_size) * prior_mean))
        return (updated_mean, total_samples)


def var_batch(new_batch, prior_mean:float=None, prior_sum_squares:float=None, prior_sample_size:int=None):
    """
    Find the new (approximate) variance of a sample updated by one batch.
    If only `new_batch` is supplied, `np.var` is used.

    Parameters
    ----------
        new_batch: List[Union[int, float]]
            List of all values in the new batch
        prior_mean: float
            Mean up until the new batch.
        prior_sum_squares: float
            Sum of the squares of the prior batch.
        prior_sample_size: int
            Number of samples in the prior batches.
    
    Returns
    -------
        tuple(float, float)
            1. The variance of the new batch and prior batches.
            2. The sum of square deviations of the new batch and prior batches.

    Examples
    --------
    >>> from stats_batch import var_batch
    >>> var_batch([1,2,3,4])
    (1.25, 5.0)
    >>> var_batch([1,2,3,4], prior_mean = 0, prior_sum_squares = 5, prior_sample_size = 4)
    (3.2142857142857144, 22.5)

    References
    ----------
    See Gandrud (2021) <https://elegant-heyrovsky-54a43f.netlify.app/privacy-first-ds-mean-var.html> for algorithm details.

    This work is based on:
    - Chou (2021, 5) <https://arxiv.org/pdf/2102.03316.pdf>

    - Chan et al (1983) <http://www.cs.yale.edu/publications/techreports/tr222.pdf> 
    """
    if prior_sum_squares is None or prior_mean is None or prior_sample_size is None:
        return (np.var(new_batch), sum_square_deviations(new_batch))
    else:
        batch_mean = np.mean(new_batch)
        total_samples = prior_sample_size + len(new_batch)
        ssd_new_batch = sum_square_deviations(new_batch)
        new_ssd = (prior_sum_squares + ssd_new_batch) + \
            (prior_sample_size / total_samples) * \
            (total_samples - prior_sample_size) * \
            (batch_mean - prior_mean) ** 2
        var_new = new_ssd / (total_samples - 1)
        return (var_new, new_ssd)

class MeanVarBatch:
    """
    Class for mean and variance of a sample created through batch updating.
    """
    def __init__(self, mean, var, sum_squares, sample_size):
        self.mean = mean
        self.var = var
        self.sum_squares = sum_squares
        self.sample_size = sample_size

    def __getitem__(self, item):
        return getattr(self, item)

    def to_pandas(self):
        """
        Returns a pandas dataframe with the mean, variance, sum of squared deviations 
        and sample size.

        Returns
        -------
        pandas.DataFrame

        Examples
        --------
        >>> sb.mean_var_batch([1,2,3,4]).to_pandas()
           mean   var  sum_squared_dev  sample_size
        0   2.5  1.25              5.0            4
        """
        return pd.DataFrame({"mean": [self.mean], "var": [self.var], 
        "sum_squared_dev": [self.sum_squares], "sample_size": [self.sample_size]})

    def print(self):
        return self.to_pandas()

    def update(self, new_batch):
        """
        Updates the existing object with a new batch.

        Parameters
        ----------
        new_batch: List[Union[int, float]]
            List of all values in the new batch
        """
        updated = mean_var_batch(new_batch, self.mean, self.sum_squares, self.sample_size)
        self.__dict__.update(updated.__dict__)

    def ttest_ind(self, other):
        """
        T-test for means of two independent samples

        Parameters
        ----------
        other : MeanVarBatch
            The other sample's sufficient statistics as generated by `mean_var_batch` 

        Returns
        -------
        (float, float)
            The t-statistic and P-value.
        """
        return ttest_ind_from_stats(mean1=self.mean, mean2=other.mean,
                                   std1=sqrt(self.var), std2=sqrt(other.var),
                                   nobs1=self.sample_size, nobs2=other.sample_size)    

    def to_csv(self, filename):
        """
        Saves mean, variance, sum of squared deviations, and sample size to a csv file.

        Parameters
        ----------
        filename : str
            The name of the file to save.

        Examples
        --------
        >>> sb.mean_var_batch([1,2,3,4]).to_csv("test.csv")
        """
        self.to_pandas().to_csv(filename)

def mean_var_batch(new_batch, prior_mean:float=None, prior_sum_squares:float=None, prior_sample_size:int=None):
    """
    Find the new (approximate) mean and variance of a sample updated by one batch.
    If only `new_batch` is supplied, `np.mean` and `np.var` are used.

    Parameters
    ----------
        new_batch: List[Union[int, float]]
            List of all values in the new batch
        prior_mean: float
            Mean up until the new batch.
        prior_sum_squares: float
            Sum of the squares of the prior batch.
        prior_sample_size: int
            Number of samples in the prior batches.
    
    Returns
    -------
        tuple(float, float, float, int) 
        1. The mean of the new batch and prior batches.
        2. The variance of the new batch and prior batches.
        3. The sum of square deviations of the new batch and prior batches.
        4. The sample size of the new batch and prior batches.

    Examples
    --------
    >>> from stats_batch import mean_var_batch
    >>> mean_var_batch([1,2,3,4])
    >>> mean_var_batch([1,2,3,4], prior_mean = 2.5, prior_sum_squares = 5, prior_sample_size = 4)
    """
    b_mean, b_n = mean_batch(new_batch, prior_mean, prior_sample_size)
    b_var, b_ssd = var_batch(new_batch, prior_mean, prior_sum_squares, prior_sample_size)
    return  MeanVarBatch(b_mean, b_var, b_ssd, b_n)

        